### PR TITLE
Refactor file writing in OSRM contract

### PR DESCRIPTION
**Pull request by TheMarex**
# Issue

This works towards #3899 by refactoring the file code in the contractor.

## Tasklist
 - [x] add regression / cucumber cases (see docs/testing.md)
 - [x] review
 - [x] adjust for comments

### CODE REVIEW

**Comment by TheMarex** [comment]
Urg some tetss are failing need to investigate.

File: include/contractor/graph_contractor.hpp [hunk 1]
```diff
@@ -101,12 +101,14 @@ class GraphContractor
 
     void Run(double core_factor = 1.0);
 
-    void GetCoreMarker(std::vector<bool> &out_is_core_node);
+    std::vector<bool> GetCoreMarker();
```
Comment (MoKob) [inline include/contractor/graph_contractor.hpp]: Thanks for getting rid of the `please fill my vector` way. I feel this way is much better. How do you feel about adding some extra commentary here? As far as I read the code, both vectors get destroyed when accessing here, right?

File: include/util/deallocating_vector.hpp [hunk 2]
```diff
@@ -258,8 +258,12 @@ template <typename ElementT> class DeallocatingVector
     DeallocatingVector(DeallocatingVector &other) = delete;
     DeallocatingVector &operator=(DeallocatingVector &other) = delete;
     // moving is fine
-    DeallocatingVector(DeallocatingVector &&other) = default;
-    DeallocatingVector &operator=(DeallocatingVector &&other) = default;
+    DeallocatingVector(DeallocatingVector &&other) { swap(other); }
+    DeallocatingVector &operator=(DeallocatingVector &&other)
```
Comment (MoKob) [inline include/util/deallocating_vector.hpp]: I know where this came from, still I am a bit confused on why this change was necessary. Was it just unexplainable behaviour, or did you find some reason for the problem?

File: include/util/vector_view.hpp [hunk 3]
```diff
@@ -143,8 +143,33 @@ template <> class vector_view<bool>
     unsigned *m_ptr;
     std::size_t m_size;
 
+    static constexpr std::size_t UNSIGNED_BITS = CHAR_BIT * sizeof(unsigned);
```
Comment (MoKob) [inline include/util/vector_view.hpp]: 👍

File: include/util/vector_view.hpp [hunk 4]
```diff
@@ -143,8 +143,33 @@ template <> class vector_view<bool>
...
+
   public:
     using value_type = bool;
+    struct reference
```
Comment (MoKob) [inline include/util/vector_view.hpp]: Nitpick/Questions: I am not entirely sure on the substructures naming convention, so feel free to ignore: shouldn't this be capital case?

File: include/util/vector_view.hpp [hunk 5]
```diff
@@ -166,6 +191,13 @@ template <> class vector_view<bool>
 
     bool operator[](const unsigned index) const { return at(index); }
 
+    reference operator[](const unsigned index)
+    {
+        const std::size_t bucket = index / UNSIGNED_BITS;
```
Comment (MoKob) [inline include/util/vector_view.hpp]: Opposed to `at`, this one is not asserting the index anymore. Could you add an assert like in `at`?

File: include/util/deallocating_vector.hpp (hunk 2 above)
Comment (TheMarex) [inline include/util/deallocating_vector.hpp]: It basically caused a memory leak because `buckets = std::move(other.buckets)` (the auto generated code) would not free the buckets.

File: include/util/vector_view.hpp (hunk 4 above)
Comment (TheMarex) [inline include/util/vector_view.hpp]: This needs to be lower-case to be compatible with the vector naming of this type.

**Comment by TheMarex** [comment]
Addressed the PR comments, just waiting on 💚 before merge.
